### PR TITLE
Minor fixes for circuits.

### PR DESCRIPTION
**Pull request by d0cd**
This PR, 
- implements the optimization in #1348 
- fixes a check in `LinearCombination::is_boolean_type`

### CODE REVIEW

File: circuit/types/boolean/src/not.rs [hunk 1]
```diff
@@ -38,6 +38,8 @@ impl<E: Environment> Not for &Boolean<E> {
             // Constant case.
             true => Boolean(E::one() - &self.0),
             // Public and private cases.
+            // Note: We must directly instantiate a public variable due to the way we represent a boolean in a linear combination.
+            // For more information, see `LinearCombination::is_boolean_type`.
             false => Boolean(Variable::Public(0, Rc::new(E::BaseField::one())) - &self.0),
```
Comment (Pratyush) [inline circuit/types/boolean/src/not.rs:43]: Can this not be `E::one()`?

Comment (Pratyush) [reply circuit/types/boolean/src/not.rs:43]: i.e.

```suggestion
            false => Boolean(E::one()),
```

File: circuit/types/integers/src/helpers/from_bits.rs [hunk 2]
```diff
@@ -24,11 +24,10 @@ impl<E: Environment, I: IntegerType> FromBits for Integer<E, I> {
         // Ensure the list of booleans is within the allowed size in bits.
         let num_bits = bits_le.len() as u64;
         if num_bits > I::BITS {
-            // Check if all excess bits are zero.
-            let should_be_zero =
-                bits_le[I::BITS as usize..].iter().fold(Boolean::constant(false), |acc, bit| acc | bit);
-            // Ensure `should_be_zero` is zero.
-            E::assert_eq(E::zero(), should_be_zero);
+            // Check that all excess bits are zero.
+            for bit in bits_le[I::BITS as usize..].iter() {
```
Comment (Pratyush) [inline circuit/types/integers/src/helpers/from_bits.rs]:
```suggestion
            for bit in &bits_le[I::BITS as usize..] {
```

File: circuit/types/scalar/src/helpers/from_bits.rs [hunk 3]
```diff
@@ -35,10 +35,10 @@ impl<E: Environment> FromBits for Scalar<E> {
         // Ensure the list of booleans is within the allowed size in bits.
         let num_bits = bits_le.len();
         if num_bits > size_in_bits {
-            // Check if all excess bits are zero.
-            let should_be_zero = bits_le[size_in_bits..].iter().fold(Boolean::constant(false), |acc, bit| acc | bit);
-            // Ensure `should_be_zero` is zero.
-            E::assert_eq(E::zero(), should_be_zero);
+            // Check that all excess bits are zero.
+            for bit in bits_le[size_in_bits..].iter() {
+                E::assert_eq(E::zero(), bit);
```
Comment (Pratyush) [inline circuit/types/scalar/src/helpers/from_bits.rs]:
```suggestion
            for bit in &bits_le[size_in_bits..] {
                E::assert_eq(E::zero(), bit);
```

File: circuit/types/boolean/src/not.rs (hunk 1 above)
Comment (d0cd) [reply circuit/types/boolean/src/not.rs:43]: It produces an error saying that the boolean is not well formed.
```
Both LC::constant and LC::terms contain elements, which is a violation
thread 'not::tests::test_not_private' panicked at 'Boolean variable is not well-formed', []/snarkVM/circuit/environment/src/circuit.rs:246:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Comment (d0cd) [reply circuit/types/boolean/src/not.rs:43]: Related, what's the intuition for representing boolean values in this way?
https://github.com/AleoHQ/snarkVM/blob/0c20d9ae0d96b0d8704e85ce682551c43153b072/circuit/environment/src/helpers/linear_combination.rs#L91
Is it to ensure that different modes have distinct representations?

Comment (d0cd) [reply circuit/types/boolean/src/not.rs:43]: This representation is a compact way to represent boolean values, with efficient operations.

**Review by howardwu — APPROVED** [review]
LGTM

**Pull request merged**